main.go

```diff
@@ -585,27 +476,64 @@ type metrics struct {
 	hasFoundResults bool
 }
 
-func runSingleScan(ctx context.Context, cfg scanConfig, scanEntireChunk bool) (metrics, error) {
-	eng, err := engine.Start(ctx,
-		engine.WithConcurrency(cfg.Concurrency),
-		engine.WithDecoders(cfg.Decoders...),
-		engine.WithDetectors(engine.DefaultDetectors()...),
-		engine.WithDetectors(cfg.Conf.Detectors...),
-		engine.WithVerify(!cfg.NoVerification),
-		engine.WithFilterDetectors(cfg.IncludeFilter),
-		engine.WithFilterDetectors(cfg.ExcludeFilter),
-		engine.WithFilterDetectors(cfg.EndpointCustomizer),
-		engine.WithFilterUnverified(cfg.FilterUnverified),
-		engine.WithResults(cfg.ParsedResults),
-		engine.WithPrintAvgDetectorTime(cfg.PrintAvgDetectorTime),
-		engine.WithPrinter(cfg.Printer),
-		engine.WithFilterEntropy(cfg.FilterEntropy),
-		engine.WithVerificationOverlap(cfg.AllowVerificationOverlap),
-		engine.WithEntireChunkScan(scanEntireChunk),
-	)
+func runSingleScan(ctx context.Context, cmd string, cfg engine.Config) (metrics, error) {
+	var scanMetrics metrics
+
+	// Setup job report writer if provided
+	var jobReportWriter io.WriteCloser
+	if *jobReportFile != nil {
+		jobReportWriter = *jobReportFile
+	}
+
+	handleFinishedMetrics := func(ctx context.Context, finishedMetrics <-chan sources.UnitMetrics, jobReportWriter io.WriteCloser) {
+		go func() {
+			defer func() {
+				jobReportWriter.Close()
+				if namer, ok := jobReportWriter.(interface{ Name() string }); ok {
+					ctx.Logger().Info("report written", "path", namer.Name())
+				} else {
+					ctx.Logger().Info("report written")
+				}
+			}()
+
+			for metrics := range finishedMetrics {
+				metrics.Errors = common.ExportErrors(metrics.Errors...)
+				details, err := json.Marshal(map[string]any{
+					"version": 1,
+					"data":    metrics,
+				})
+				if err != nil {
+					ctx.Logger().Error(err, "error marshalling job details")
+					continue
+				}
+				if _, err := jobReportWriter.Write(append(details, '\n')); err != nil {
+					ctx.Logger().Error(err, "error writing to file")
+				}
+			}
+		}()
+	}
+
+	const defaultOutputBufferSize = 64
+	opts := []func(*sources.SourceManager){
+		sources.WithConcurrentSources(cfg.Concurrency),
+		sources.WithConcurrentUnits(cfg.Concurrency),
+		sources.WithSourceUnits(),
+		sources.WithBufferedOutput(defaultOutputBufferSize),
+	}
+
+	if jobReportWriter != nil {
+		unitHook, finishedMetrics := sources.NewUnitHook(ctx)
+		opts = append(opts, sources.WithReportHook(unitHook))
```

Comment:
@mcastorina Since the `sourceManager` is responsible for closing the `finishedMetrics` channel, this should still work, correct? The `opts` would be used when initializing the `sourceManager`, which then gets passed into the engine.